### PR TITLE
Shortest path partitioning

### PR DESCRIPTION
**Pull request by buckysballs**

### CODE REVIEW

File: src/main/scala/org/constellation/trust/DataGeneration.scala [hunk 1]
```diff
@@ -33,30 +34,57 @@ case class TrustNode(id: Int, xCoordinate: Double, yCoordinate: Double, edges: S
...
+  def positiveEdges: Seq[TrustEdge] = edges.filter(_.trust > 0)
+
 }
 
 object DataGeneration {
   val sqrt2: Double = Math.sqrt(2)
 
-  def generateTestData(numNodes: Int = 30): List[TrustNode] = {
+  def randomEdgeLogic(random: Random = new Random())(distance: Double): Boolean =
+    Random.nextDouble() > distance && Random.nextDouble() < 0.5
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/trust/DataGeneration.scala:45]: Random is a side effect. It should be in an effect type F (IO)

File: src/main/scala/org/constellation/trust/DataGeneration.scala [hunk 2]
```diff
@@ -33,30 +34,57 @@ case class TrustNode(id: Int, xCoordinate: Double, yCoordinate: Double, edges: S
...
+
+  def randomEdge(random: Random = new Random())(logic: Double => Boolean = randomEdgeLogic(random))(n: TrustNode, n2: TrustNode) =
+    if (logic(n.id)) {
+      val trustZeroToOne = Random.nextDouble()
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/trust/DataGeneration.scala:49]: as above

File: src/main/scala/org/constellation/trust/DataGeneration.scala [hunk 3]
```diff
@@ -33,30 +34,57 @@ case class TrustNode(id: Int, xCoordinate: Double, yCoordinate: Double, edges: S
...
+      Some(TrustEdge(n.id, n2.id, 2 * (trustZeroToOne - 0.5), isLabel = true))
+    } else None
+
+  def generateFullyConnectedTestData(numNodes: Int = 30) =
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/trust/DataGeneration.scala:53]: Test data should be generated in tests only. It shouldn't be mixed with the protocol code.

**Review by marcinwadon — COMMENTED** [review]
Looks really good, especially usage of cats and TopKMonoid :)

File: src/main/scala/org/constellation/trust/DataGeneration.scala [hunk 4]
```diff
@@ -33,30 +34,57 @@ case class TrustNode(id: Int, xCoordinate: Double, yCoordinate: Double, edges: S
...
+    generateTestData(numNodes, cliqueEdge()(seedCliqueLogic(numNodes)))
+
+  def seedCliqueLogic(maxSeedNodeIdx: Int = 1)(id: Double): Boolean =
```
[the remark this answers is not in the document]
Comment (buckysballs) [reply src/main/scala/org/constellation/trust/DataGeneration.scala:56]: We need the false condition or its seen as AnyVal

[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/trust/DataGeneration.scala:56]: @buckysballs: @kpudlik made a typo, it should be:
```
def seedCliqueLogic(maxSeedNodeIdx: Int = 1)(id: Double): Boolean =
  id <= maxSeedNodeIdx
```

**Comment by marcinwadon** [comment]
@buckysballs Have you tested the partitioner with different reputation across the nodes/participants?

**Review by kpudlik — DISMISSED** [review]

**Pull request closed without merging**